### PR TITLE
Safe error handling for command API

### PR DESCRIPTION
**Pull request by GabrielMajeri**
**Connections**
Part of #638 

**Description**
Initially started to remove assertions from the `bundle.rs` file, but ended up implementing safe error handling for the whole `command` module.

**Testing**
Checked with core and player.

### CODE REVIEW

File: wgpu-core/src/command/bundle.rs [hunk 1]
```diff
@@ -62,9 +63,9 @@ impl RenderBundleEncoder {
         desc: &wgt::RenderBundleEncoderDescriptor,
         parent_id: id::DeviceId,
         base: Option<BasePass<RenderCommand>>,
-    ) -> Self {
+    ) -> Result<Self, InvalidSampleCountError> {
```
Comment (kvark) [inline wgpu-core/src/command/bundle.rs:66]: let's turn this error into
```rust
enum CreateRenderBundleError {
  InvalidSampleCount(u8),
}
```

File: wgpu-core/src/command/bundle.rs [hunk 2]
```diff
@@ -579,13 +588,50 @@ impl State {
...
+    /// - The pipeline is not compatible with the depth-stencil read-only render pass.
+    IncompatiblePipeline,
+    /// The provided buffer does not have the required usage.
+    InvalidBufferUsage,
+    /// The provided texture does not have the required usage.
+    InvalidTextureUsage,
+    /// A pipeline must be bound.
+    UnboundPipeline,
+    /// There was an error with the push constants.
+    PushConstantUploadError(PushConstantUploadError),
```
Comment (kvark) [inline wgpu-core/src/command/bundle.rs:613]: maybe just `PushConstants(PushConstantUploadError)`?

File: wgpu-core/src/command/bundle.rs [hunk 3]
```diff
@@ -621,32 +667,30 @@ impl<G: GlobalIdentityHandlerFactory> Global<G> {
...
-                                off,
-                                wgt::BIND_BUFFER_ALIGNMENT
-                            );
+                        // Check for misaligned offsets.
+                        if offsets
+                            .iter()
+                            .map(|offset| *offset as wgt::BufferAddress)
+                            .any(|offset| offset % wgt::BIND_BUFFER_ALIGNMENT != 0)
+                        {
+                            return Err(RenderCommandError::UnalignedBufferOffset);
```
Comment (kvark) [inline wgpu-core/src/command/bundle.rs:683]: maybe include the offset in the error?

File: wgpu-core/src/command/bundle.rs [hunk 4]
```diff
@@ -621,32 +667,30 @@ impl<G: GlobalIdentityHandlerFactory> Global<G> {
...
                         }
 
                         let bind_group = state
                             .trackers
                             .bind_groups
                             .use_extend(&*bind_group_guard, bind_group_id, (), ())
                             .unwrap();
-                        assert_eq!(bind_group.dynamic_binding_info.len(), offsets.len());
+                        if bind_group.dynamic_binding_info.len() != offsets.len() {
+                            return Err(RenderCommandError::WrongBufferOffsetCount);
```
Comment (kvark) [inline wgpu-core/src/command/bundle.rs:692]: Maybe `InvalidDynamicOffsetCount`?

File: wgpu-core/src/command/bundle.rs [hunk 5]
```diff
@@ -688,7 +731,9 @@ impl<G: GlobalIdentityHandlerFactory> Global<G> {
                             .buffers
                             .use_extend(&*buffer_guard, buffer_id, (), BufferUse::INDEX)
                             .unwrap();
-                        assert!(buffer.usage.contains(wgt::BufferUsage::INDEX), "An invalid setIndexBuffer call has been made. The buffer usage is {:?} which does not contain required usage INDEX", buffer.usage);
+                        if !buffer.usage.contains(wgt::BufferUsage::INDEX) {
+                            return Err(RenderCommandError::InvalidBufferUsage);
```
Comment (kvark) [inline wgpu-core/src/command/bundle.rs:735]: Maybe `MissingBufferUsage(wgt::BufferUsage::INDEX)`?

File: wgpu-core/src/command/compute.rs [hunk 6]
```diff
@@ -107,14 +108,42 @@ struct State {
...
+    /// A pipeline must be bound.
+    UnboundPipeline,
+    /// The provided buffer does not have the required usage.
+    InvalidBufferUsage,
+    /// Can't pop debug group, because number of pushed debug groups is zero.
+    InvalidPopDebugGroup,
+    /// An error encountered while validating dynamic bindings.
+    BindError(BindError),
+    /// There was an error with the push constants.
+    PushConstantUploadError(PushConstantUploadError),
```
Comment (kvark) [inline wgpu-core/src/command/compute.rs:124]: same note here

File: wgpu-core/src/command/compute.rs [hunk 7]
```diff
@@ -162,12 +191,9 @@ impl<G: GlobalIdentityHandlerFactory> Global<G> {
...
                     bind_group_id,
                 } => {
-                    assert!(
-                        (index as u32) < cmb.limits.max_bind_groups,
-                        "Bind group index {0} is out of range 0..{1} provided by requested max_bind_group limit {1}",
-                        index,
-                        cmb.limits.max_bind_groups
-                    );
+                    if (index as u32) >= cmb.limits.max_bind_groups {
+                        return Err(ComputePassError::BindGroupIndexOutOfRange);
```
Comment (kvark) [inline wgpu-core/src/command/compute.rs:195]: let's include the index

File: wgpu-core/src/command/compute.rs [hunk 8]
```diff
@@ -294,37 +320,34 @@ impl<G: GlobalIdentityHandlerFactory> Global<G> {
...
+                    let pipeline_layout = &pipeline_layout_guard[pipeline_layout_id];
 
                     pipeline_layout
                         .validate_push_constant_ranges(
                             wgt::ShaderStage::COMPUTE,
                             offset,
                             end_offset_bytes,
                         )
-                        .unwrap();
+                        .map_err(ComputePassError::from)?;
```
Comment (kvark) [inline wgpu-core/src/command/compute.rs:335]: You don't actually need this map, the `?` operation should do it for you

File: wgpu-core/src/command/render.rs [hunk 9]
```diff
@@ -38,24 +39,20 @@ pub type DepthStencilAttachmentDescriptor =
...
     desc: &DepthStencilAttachmentDescriptor,
     aspects: hal::format::Aspects,
-) -> bool {
+) -> Result<bool, RenderPassError> {
     if aspects.contains(hal::format::Aspects::DEPTH) && !desc.depth.read_only {
-        return false;
+        return Ok(false);
+    }
+    if (desc.depth.load_op, desc.depth.store_op) != (LoadOp::Load, StoreOp::Store) {
+        return Err(RenderPassError::InvalidDepthStencilOps);
```
Comment (kvark) [inline wgpu-core/src/command/render.rs:47]: let's separate this into `InvalidDepthOps` and `InvalidStencilOps`

File: wgpu-core/src/command/render.rs [hunk 10]
```diff
@@ -346,14 +343,67 @@ impl State {
...
+    /// Indirect draw command would overrun the bounds of the buffer.
+    IndirectBufferOverrun,
+    /// Can't pop debug group, because number of pushed debug groups is zero.
+    InvalidPopDebugGroup,
+    /// The render bundle output formats do not match render pass attachment formats.
+    IncompatibleRenderBundle,
+    /// An error encountered while encoding a command.
+    RenderCommandError(RenderCommandError),
+    /// An error encountered while handling a draw command.
+    DrawError(DrawError),
```
Comment (kvark) [inline wgpu-core/src/command/render.rs:376]: similarly here and in other places, we don't need `Error` suffix in variants inside an `Error` kind of enum.
This could just be `Draw(DrawError)` and `Bind(BindError)`

File: wgpu-core/src/command/compute.rs [hunk 11]
```diff
@@ -107,14 +108,42 @@ struct State {
...
+    /// A bind group index is greater than the device's requested `max_bind_group` limit.
+    BindGroupIndexOutOfRange,
+    /// A pipeline must be bound.
+    UnboundPipeline,
+    /// The provided buffer does not have the required usage.
+    InvalidBufferUsage,
+    /// Can't pop debug group, because number of pushed debug groups is zero.
+    InvalidPopDebugGroup,
+    /// An error encountered while validating dynamic bindings.
+    BindError(BindError),
```
Comment (kvark) [inline wgpu-core/src/command/compute.rs:122]: let's rename to just `Bind(BindError)`

File: wgpu-core/src/command/render.rs [hunk 12]
```diff
@@ -896,7 +938,9 @@ impl<G: GlobalIdentityHandlerFactory> Global<G> {
                         .bind_groups
                         .use_extend(&*bind_group_guard, bind_group_id, (), ())
                         .unwrap();
-                    bind_group.validate_dynamic_bindings(offsets).unwrap();
+                    bind_group
+                        .validate_dynamic_bindings(offsets)
+                        .map_err(RenderPassError::from)?;
```
Comment (kvark) [inline wgpu-core/src/command/render.rs:943]: similarly here, we don't need this `map_err()`, just do `validate_dynamic_bindings(offsets)?`

File: wgpu-core/src/command/render.rs [hunk 13]
```diff
@@ -1251,53 +1289,35 @@ impl<G: GlobalIdentityHandlerFactory> Global<G> {
...
                     };
 
                     if count.is_some() {
-                        assert!(
-                            device.features.contains(wgt::Features::MULTI_DRAW_INDIRECT),
-                            "The feature MULTI_DRAW_INDIRECT must be enabled to use {}",
-                            name
-                        );
+                        if !device.features.contains(wgt::Features::MULTI_DRAW_INDIRECT) {
+                            return Err(RenderPassError::MissingDeviceFeature(
```
Comment (kvark) [inline wgpu-core/src/command/render.rs:1301]: let's use the actual bitflags in the error instead of string slices. Users can always convert them to strings if they need that.

File: wgpu-core/src/command/bundle.rs (hunk 5 above)
Comment (cwfitzgerald) [reply wgpu-core/src/command/bundle.rs:735]: For things that have a log message, I would prefer that the original message gets `log::error!`'d to the logs so that users can still see the full message written in plain english.

File: wgpu-core/src/command/compute.rs (hunk 8 above)
Comment (GabrielMajeri) [reply wgpu-core/src/command/compute.rs:362]: It does, but here we'd need a double `.into`, so `?` alone won't work.

**Comment by GabrielMajeri** [comment]
I've updated the PR, now all error types have `Display` implementations which provide detailed error messages.
